src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptInProcLanguageClient.cs

```diff
@@ -0,0 +1,101 @@
+            [Import(AllowDefault = true)] IVSTypeScriptCapabilitiesProvider? typeScriptCapabilitiesProvider,
+            VSTypeScriptRequestDispatcherFactory requestDispatcherFactory,
+            IGlobalOptionService globalOptions,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            LspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+            DefaultCapabilitiesProvider defaultCapabilitiesProvider,
+            ILspLoggerFactory lspLoggerFactory,
+            IThreadingContext threadingContext)
+            : base(requestDispatcherFactory, globalOptions, listenerProvider, lspWorkspaceRegistrationService, lspLoggerFactory, threadingContext, diagnosticsClientName: null)
+        {
+            _typeScriptCapabilitiesProvider = typeScriptCapabilitiesProvider;
+        }
+
+        protected override ImmutableArray<string> SupportedLanguages => ImmutableArray.Create(InternalLanguageNames.TypeScript);
+
+        public override ServerCapabilities GetCapabilities(ClientCapabilities clientCapabilities)
+        {
+            var serverCapabilities = GetTypeScriptServerCapabilities(clientCapabilities);
+
+            serverCapabilities.TextDocumentSync = new TextDocumentSyncOptions
+            {
+                Change = TextDocumentSyncKind.Incremental,
+                OpenClose = true,
+            };
+
+            serverCapabilities.ProjectContextProvider = true;
+
+            var isPullDiagnostics = GlobalOptions.IsPullDiagnostics(InternalDiagnosticsOptions.NormalDiagnosticMode);
+            if (isPullDiagnostics)
+            {
+                serverCapabilities.SupportsDiagnosticRequests = true;
+            }
+
+            return serverCapabilities;
+        }
+
+        /// <summary>
+        /// When pull diagnostics is enabled, ensure that initialization failures are displayed to the user as
+        /// they will get no diagnostics.  When not enabled we don't show the failure box (failure will still be recorded in the task status center)
+        /// as the failure is not catastrophic.
+        /// </summary>
+        public override bool ShowNotificationOnInitializeFailed => GlobalOptions.IsPullDiagnostics(InternalDiagnosticsOptions.NormalDiagnosticMode);
+
+        public override WellKnownLspServerKinds ServerKind => WellKnownLspServerKinds.RoslynTypeScriptLspServer;
+
+        private VSInternalServerCapabilities GetTypeScriptServerCapabilities(ClientCapabilities clientCapabilities)
+        {
+            if (_typeScriptCapabilitiesProvider != null)
+            {
+                var serializedClientCapabilities = JsonConvert.SerializeObject(clientCapabilities);
+                var serializedServerCapabilities = _typeScriptCapabilitiesProvider.GetServerCapabilities(serializedClientCapabilities);
+                var typeScriptServerCapabilities = JsonConvert.DeserializeObject<VSInternalServerCapabilities>(serializedServerCapabilities);
```

Comment:
talk in terms of serialized capabilities to avoid hard dependencies on protocol types